PSDP: Woodgate Algorithm

banrovegrie opened the pull request:
The basic algorithm has been implemented. Reformatting the code and testing are left. Putting in this PR so that we can discuss this tomorrow.

banrovegrie commented:
@FanwangM let's run tests only after I have reformatted the code.

banrovegrie commented:
I tested the algorithm for square matrices and it appears to work. 3 doubts remain which I have posted in the discussion forum.

banrovegrie commented:
@FanwangM could you help me out on where tox is failing?

FanwangM commented on procrustes/woodgate.py:
```diff
@@ -0,0 +1,260 @@
+        E = E - w_min * D
+        i += 1
+
+    print(f"Woodgate's algorithm took {i} iterations.")
+    print(f"Error = {np.linalg.norm(F - E.T @ E @ G)}.")
+    print(f"Required P = {E.T @ E}")
```
We don't need this information in the final version and it's safe to comment them out.

banrovegrie replied:
removed in 3f938e10c7d2069dbcced45d25a57b500bf60de1

FanwangM commented on procrustes/woodgate.py:
```diff
@@ -0,0 +1,260 @@
+    print(f"Woodgate's algorithm took {i} iterations.")
+    print(f"Error = {np.linalg.norm(F - E.T @ E @ G)}.")
+    print(f"Required P = {E.T @ E}")
+    return E.T @ E
```
We will return a `ProcrustesResult` object in the end.

FanwangM commented:
Can you help finalize this algorithm as we discussed including the formatting style?
Please feel free to tag me if you have questions. Thank you. @banrovegrie

banrovegrie commented on procrustes/woodgate.py:
```diff
@@ -0,0 +1,260 @@
+    E = np.random.rand(n, n)
+
+    while True:
+        LE = L(E)
```
it has been renamed and is a lambda function at line 286

banrovegrie commented on procrustes/woodgate.py:
```diff
@@ -0,0 +1,260 @@
+        The permutation matrix.
+    """
+    k = 0
+    x, _ = arr.shape
```
fixed in 3f938e10c7d2069dbcced45d25a57b500bf60de1

banrovegrie commented on procrustes/woodgate.py:
```diff
@@ -0,0 +1,260 @@
+import numpy as np
+import scipy.linalg as lin
+from scipy.optimize import minimize
+# from procrustes.utils import ProcrustesResult
```
There is an issue with using ProcrastesResult, I will elaborate below,

banrovegrie commented:
@FanwangM The algorithm has exact precision as expected. After implementing scaling and choosing $E_0 = I_n$, I obtained the exact results mentioned in the paper's examples.

Our algorithm also converges equally as fast.

banrovegrie commented on procrustes/woodgate.py:
```diff
@@ -0,0 +1,260 @@
+__all__ = ["woodgate"]
+
+
+def permutation_matrix(arr: np.ndarray) -> np.ndarray:
```
Oh this method is not related to the permutation Procrustes problem. It just deals with finding a matrix such that $v(E^T') = Pv(E)$.

Here, $v$ represents flattenning.

banrovegrie commented on procrustes/woodgate.py:
```diff
@@ -0,0 +1,260 @@
+    bool
+        True if the matrix is positive semidefinite.
+    """
+    return np.all(np.linalg.eigvals(x) >= 0)
```
That is true, but won't it hamper readability within the `woodgate` function where this is method is being called.

FanwangM replied:
If you want to keep it, we need at least to make it a private function as it's only used for once and only used in this module. @banrovegrie

banrovegrie replied:
Ok, I will make it private.

banrovegrie commented on procrustes/woodgate.py:
```diff
@@ -0,0 +1,260 @@
+    eigenvalues, U = np.linalg.eig(arr)
+    eigenvalues_pos = [max(0, i) for i in eigenvalues]
+    U_inv = np.linalg.inv(U)
+    return U @ np.diag(eigenvalues_pos) @ U_inv
```
I will fix it in the next commit.

banrovegrie replied:
Fixed in https://github.com/theochem/procrustes/pull/170/commits/cee97fa4cc47bf017b24ed181f0f1d3d3e15eb24

banrovegrie commented:
Hey, so I ran into an issue while using `ProcrustesResult` to return the error and required transformation. In case of `ProcrustesResult`, we aim to return $S, T$. They are defined as follows.

$$
\|\mathbf{S}\mathbf{A}\mathbf{T} - \mathbf{B}\|_{F}^2 =
\text{Tr}\left[
     \left(\mathbf{S}\mathbf{A}\mathbf{T} - \mathbf{B}\right)^\dagger
     \left(\mathbf{S}\mathbf{A}\mathbf{T} - \mathbf{B}\right)\right]
$$

However, in our program, we are dealing with transformation only on the left-hand side. Namely, $\min_{\mathbf{P}} \|F - \mathbf{P}G\|^2$.

How should I proceed for this?

banrovegrie commented:
Also, abiding by the terminology present in other modules, $A$ is considered as the matrix on which the transformation will be applied and $B$ is the target matrix.

Thus, our result in `woodgate` is effectively finding a left hand transformation $P$ such that $\|PA - B\|^2 = \|B - PA\|^2$ is minimum. Note that to draw parallel with the paper, $A$ is $G$ and $B$ is the matrix $F$.

banrovegrie commented:
Other than these two persisting questions, everything else is complete. @FanwangM

FanwangM commented:
> Hey, so I ran into an issue while using `ProcrustesResult` to return the error and required transformation. In case of `ProcrustesResult`, we aim to return S,T. They are defined as follows.
> 
> |SAT−B|F2=Tr[(SAT−B)†(SAT−B)]
> 
> However, in our program, we are dealing with transformation only on the left-hand side. Namely, minP|F−PG|2.
> 
> How should I proceed for this?

The other matrix is defined as the identity matrix with the same shape.

FanwangM commented:
The linter keeps complaining and needs to be fixed.
Also, can you also add a test file for this module, either in this PR or another PR? Thank you.
@banrovegrie

banrovegrie commented:
Apparently, `tox` says the error is in `softassign.py`. That shouldn't be the case right?

```
************* Module procrustes.softassign
  procrustes/softassign.py:221:15: E1121: Too many positional arguments for method call (too-many-function-args)
```

FanwangM commented:
Some general comments for this implementation.

- [ ] Rename `woodgate.py` to `psdp.py`
- [ ] Rename function `woodgate` to `psdp_woodgate` as we have two more algorithms coming.
- [ ] Only put  `psdp_woodgate` as a public function that a user can access and put other functions as private functions
- [ ] Fix the matrix representation in the docstring as we use bold characters to denote matrix,  such as $\mathbf{P}$.
- [ ] Try to avoid use _@_ operations and use `np.dot` if possible, https://numpy.org/doc/stable/reference/generated/numpy.linalg.multi_dot.html. This can help make things clean and consistent
- [ ] We will need to make the argument names consistent and you can take a look at other examples in Procrustes package for reference.

Thank you for the nice work. @banrovegrie

banrovegrie commented:
[x] Renamed `woodgate.py` to `psdp.py`.
[x] Renamed function `woodgate` to `psdp_woodgate`.
[x] `psdp_woodgate` is now the only public function.
[x] Bold characters represent matrices in docstring now.

- Testing file I will add in a different PR so that there we can talk about the efficiency of the implementation as well.
- Since, we will use the same testing for multiple algorithms, I think this makes more sense.
- In the case of small equations, I tried to use `np.dot` rather than `@` but some of the formulae are quite elaborate and converting them too to `np.dot` might make the readability hard.

banrovegrie commented:
I tested with `tox` locally. All the tests seem to pass. @FanwangM

FanwangM reviewed (changes requested):
The codes seem good for now. A few minor comments about single underscore and double underscore and we should use single underscore for all the private functions.

And we should move `psdp_woodgate` function to the top of all the private functions.
Thanks for the nice work.

banrovegrie commented on procrustes/psdp.py:
```diff
@@ -0,0 +1,330 @@
+__all__ = ["psdp_woodgate"]
+
+
+def __permutation_matrix(arr: np.ndarray) -> np.ndarray:
```
Yea true, since we don't have classes just the conventional single `_` will work.

banrovegrie replied:
fixed in https://github.com/theochem/procrustes/pull/170/commits/6f14e45f364a407bfd3ee1b36ccc89da0639e7f3

banrovegrie commented:
Oh okay you wanted me to implement `np.linalg.multi_dot` makes sense. Will do it right away.

banrovegrie commented on procrustes/psdp.py:
```diff
@@ -0,0 +1,330 @@
+    return p
+
+
+def __no_update(error: List[int], threshold: int = 1e-5) -> bool:
```
fixed in https://github.com/theochem/procrustes/pull/170/commits/6f14e45f364a407bfd3ee1b36ccc89da0639e7f3

banrovegrie commented on procrustes/psdp.py:
```diff
@@ -0,0 +1,330 @@
+    return abs(error[-1] - error[-2]) < threshold
+
+
+def __is_pos_semi_def(arr: np.ndarray) -> bool:
```
fixed in https://github.com/theochem/procrustes/pull/170/commits/6f14e45f364a407bfd3ee1b36ccc89da0639e7f3

banrovegrie commented on procrustes/psdp.py:
```diff
@@ -0,0 +1,330 @@
+    bool
+        True if the error has reduced/changed.
+    """
+    return abs(error[-1] - error[-2]) < threshold
```
fixed in https://github.com/theochem/procrustes/pull/170/commits/6f14e45f364a407bfd3ee1b36ccc89da0639e7f3

banrovegrie commented on procrustes/psdp.py:
```diff
@@ -0,0 +1,330 @@
+    """
+    n = e.shape[0]
+    s = np.linalg.matrix_rank(e)
+    v = lin.null_space(le.T).flatten()
```
fixed in https://github.com/theochem/procrustes/pull/170/commits/6f14e45f364a407bfd3ee1b36ccc89da0639e7f3

banrovegrie commented:
I am just left to implement `multi_dot`. Once that is done, could you merge it so that I can start pushing the test using another PR and simultaneously start working on the new algorithm?

banrovegrie commented:
In https://github.com/theochem/procrustes/pull/170/commits/2baf69adc757f7f8163c473c360655d3d12b0555, I have reimplemented `@` with `np.dot` or `np.linalg.mult_dot` as required.

banrovegrie commented:
@FanwangM could you run tests whenever you are free?
